Model/DutyCalculator.php

```diff
@@ -116,84 +116,333 @@ public function __construct(
+        //or whether applying duty is a must for that country or not (a setting in magento admin
+        //panel)
+        //filling out both Quote and session state as appropriate
+        //DHL quoteID is already saved into session; so not doing it here again
+        $quote = $this->checkoutSession->getQuote();
+        $duty_adjusted = $this->priceCurrency->round($duty); //copied over pre-existing code
+        //but are we supposed to round always?
+        //should not that be based on corresponding admin setting?
+        $this->checkoutSession->setReachDuty($duty_adjusted);
+
+        if ($apply || !$this->getIsOptional($address->getCountryId())) {
+            //checkbox selection was duty should be applied
+            //or applying duty is a must for that country
+            $baseCurrency = $this->storeManager->getStore()->getBaseCurrency();
+            $rate = $baseCurrency->getRate($baseCurrency->getCode());
+            $baseDuty = $duty_adjusted / $rate;
+            $quote->setBaseReachDuty($baseDuty);
+            $this->checkoutSession->setBaseReachDuty($baseDuty);
+            $quote->setReachDuty($duty_adjusted);
+            $this->checkoutSession->setReachDuty($duty_adjusted);
+            $quote->setDhlQuoteId($response['quoteId']);
+            $quote->setDhlBreakdown(json_encode($response['feeTotals']));
+            $this->checkoutSession->setDhlBreakdown($quote->getDhlBreakdown());
+            $this->checkoutSession->setApply(true); //checkbox selection /corresponding passed value
+            //implies that duty should be applied
+            $this->_logger->debug('Apply block immediately after DHL call');
+        } else {
+            //checkbox selection or parameter passed indicates that duty should not be applied
+            $quote->setBaseReachDuty(0);
+            $quote->setReachDuty(0);
+            $this->checkoutSession->setBaseReachDuty(0);
+            $this->checkoutSession->setReachDuty($duty_adjusted);
+            $quote->setDhlQuoteId('');
+            $this->checkoutSession->setDhlQuoteId('');
+            $quote->setDhlBreakdown('');
+            $this->checkoutSession->setDhlBreakdown('');
+            $this->checkoutSession->setApply(false);
+            $this->_logger->debug('Do not Apply block immediately after DHL call');
+        }
+
+        $quote->save();
+        $this->response->setSuccess(true);
+        $this->response->setDuty($duty_adjusted);
+        $this->response->setIsOptional($this->getIsOptional($address->getCountryId()));
+    }
+
+    /**
+     * Deals with a case when DHL API call does not return a value for duty
+     * @param Magento\Framework\App\Response $response
+     */
+    public function fillOutQuoteAndSessionOnError($response)
+    {
+        $this->response->setSuccess(false);
+        $this->response->setDuty(0);
+
+        if (isset($response['message'])) { //error message from DHL duty and tax api call?
```

Comment:
based on where this function is called I would also assume that this contains an error from the DHL duty and tax api call. I wonder if we should do different things based on the error received? Maybe this should be its own ticket to work out how to handle an error from the DHL API (for example, I wouldn't want to leave the checkbox available if we're getting big errors from DHL (400, something wrong with our API call at a fundamental level) but a 503 or something I'd be ok retrying.

Reply:
I think it should be a new ticket. 
"I wonder if we should do different things based on the error received? " - I do not know as I do not know what the original requirement asked for (as you know it is a balance between what the original requirement asked for and what is possible technically). This is the way pre-existing code was and I do not know whether I should assume they followed the requirement or not. Is there any ticket listing original requirements when that initial implementation was done? (If it is on top of your head otherwise I can look on JIRA myself assuming that is the only ticketing system we have from the beginning).  

Additionally for this iteration I can set reachDuty (in session) to zero - this would be needed when error is happening in the middle (country selection is changed and ...). If country selection for shipping address is not changed in between during the same session then checkbox would not appear due to the way the logic is (reachDuty is 0 or null at this stage ; checkbox appears when it is larger than 0).     But then I have to figure out some way to test it - either designing a mock service imitating DHL and generate error or on purpose sending request that would return error or ... which obviously would increase development effort for this ticket.

Reply:
Created ticket: https://withreach.atlassian.net/browse/MAG-76